java/ql/src/semmle/code/java/Annotation.qll

```diff
@@ -99,18 +166,69 @@ private predicate sourceAnnotValue(Annotation a, Method m, Expr val) {
 
 /** An abstract representation of language elements that can be annotated. */
 class Annotatable extends Element {
-  /** Holds if this element has an annotation. */
-  predicate hasAnnotation() { exists(Annotation a | a.getAnnotatedElement() = this) }
+  /** Holds if this element has an annotation, including inherited annotations. */
+  predicate hasAnnotation() { exists(getAnAnnotation()) }
+
+  /** Holds if this element has a declared annotation, excluding inherited annotations. */
+  predicate hasDeclaredAnnotation() { exists(getADeclaredAnnotation()) }
 
-  /** Holds if this element has the specified annotation. */
+  /**
+   * Holds if this element has the specified annotation, including inherited
+   * annotations.
+   */
   predicate hasAnnotation(string package, string name) {
     exists(AnnotationType at | at = getAnAnnotation().getType() |
       at.nestedName() = name and at.getPackage().getName() = package
     )
   }
 
-  /** Gets an annotation that applies to this element. */
-  Annotation getAnAnnotation() { result.getAnnotatedElement() = this }
+  /**
+   * Gets an annotation that applies to this element, including inherited annotations.
```

Comment:
This can't be right: it uses `getADeclaredAnnotation` which says it excludes inheritence

Reply:
That should work. Inheritance (as defined by `@Inherited`) only applies to classes, and CodeQL `Class` already overrides this predicate:
https://github.com/github/codeql/blob/f9b539e5b99bfeb72791061146377d5baa6f913e/java/ql/src/semmle/code/java/Type.qll#L632-L640

Reply:
Have added a comment clarifying this.